olp-cpp-sdk-core/include/olp/core/thread/TaskContinuation.h

```diff
@@ -0,0 +1,397 @@
+template <>
+class Continuation<void> {
+  template <typename OtherResultType>
+  friend class Continuation;
+
+ public:
+  using Result = void;
+
+  template <typename Callable>
+  Continuation<internal::AsyncResultType<Callable>> Then(Callable callable) {
+    using NewResultType = internal::AsyncResultType<Callable>;
+    using Function = internal::TypeToFunctionInput<NewResultType>;
+    return Then(std::function<void(Function)>(std::move(callable)));
+  }
+
+  Continuation() = delete;
+
+  template <typename NewType>
+  Continuation<internal::RemoveRefAndConst<NewType>> Then(
+      std::function<void(std::function<void(NewType)>)> async) {
+    return m_cont.Then(toAsyncTask({}, std::move(async)));
+  }
+
+  Continuation Then(std::function<void(std::function<void()>)> async);
+
+  using ContinuationImpl = internal::ContinuationImpl;
+  using OutResult = ContinuationImpl::OutResult;
+  using Task = ContinuationImpl::Task;
+  using Callback = ContinuationImpl::Callback;
+
+ private:
+  template <typename NewType>
+  static std::pair<ContinuationImpl::AsyncTask, ContinuationImpl::Task>
+  toAsyncTask(std::shared_ptr<ExecutionContext> context,
+              std::function<void(std::shared_ptr<ExecutionContext>,
+                                 std::function<void(NewType)>)>&& async) {
+    using NewResultType = internal::RemoveRefAndConst<NewType>;
+    auto asyncMoC = std::move(async);
+    return {[asyncMoC, context](void*, Callback callback) {
+              auto callbackMoC = std::move(callback);
+              asyncMoC(context, [callbackMoC](NewResultType input) {
+                callbackMoC(static_cast<NewResultType*>(&input));
+              });
+            },
+            [](void* input) {
+              NewResultType& in = *static_cast<NewResultType*>(input);
+              auto result = std::make_unique<NewResultType>(std::move(in));
+              return internal::make_untyped_unique(std::move(result));
+            }};
+  }
+
+  ContinuationImpl m_cont;
+};
+
+template <typename ResultType>
+class Continuation {
+  enum class ExecutionState { Initialized, Started, Finished };
+
+ public:
+  ResultType GetType() { return ResultType(); }
+
+  /// Adds an asynchronous task (one taking a callback) as the next step of the
+  /// Continuation.
+  template <typename Callable>
+  Continuation<internal::AsyncResultExecutionContextTypeFull<Callable>> Then(
+      Callable callable) {
+    using NewResultType =
+        internal::AsyncResultExecutionContextTypeFull<Callable>;
+    using Function = internal::TypeToFunctionInput<NewResultType>;
+    return Then(
+        std::function<void(ResultType, std::shared_ptr<ExecutionContext>,
+                           Function)>(std::move(callable)));
+  }
+
+  void Run() {
+    execution_state_ = ExecutionState::Started;
+    if (!finally_callback_) {
+      return;
+    }
+
+    if (m_cont.GetContext()->IsCancelled()) {
+      finally_callback_({olp::client::ApiError::Cancelled()});
```

Comment:
```suggestion
      finally_callback_(olp::client::ApiError::Cancelled());
```
Cancelled() is already returning an ApiError so no need to initialize an ApiError from an ApiError.

Reply:
done